core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -436,9 +437,20 @@ private boolean hasInjectiveImplicitCoercion(Type source, Type target, Object va
 
                     return true;
                 }
+                if (source instanceof DateType) {
+                    // Cast from TIMESTAMP WITH TIME ZONE to DATE and back to TIMESTAMP WITH TIME ZONE does not round trip, unless the value's zone is equal to session zone
+                    if (!getTimeZone(timestampWithTimeZoneType, value).equals(session.getTimeZoneKey())) {
+                        return false;
+                    }
+
+                    // Cast from DATE to TIMESTAMP WITH TIME ZONE is not monotonic when there is a forward DST change in the session zone
+                    if (!isTimestampToTimestampWithTimeZoneInjectiveAt(session.getTimeZoneKey().getZoneId(), getInstantWithTruncation(timestampWithTimeZoneType, value))) {
+                        return false;
+                    }
+
+                    return true;
```

Comment:
@findepi I'm not sure about the expected condition here. Should we verify whether all time parts of `value` are zero (e.g. `2022-11-10 00:00:00.000000`)?

Reply:
I don't think it's needes